### PR TITLE
Indirect prefetch

### PR DESCRIPTION
**Pull request by ahrens**
Based on https://github.com/openzfs/openzfs/pull/65, plus some cleanup.

### CODE REVIEW

**Comment by amotin** [comment]
I have no objections. If you like it this way -- great, just go!

**Comment by amotin** [comment]
Could you be more verbose about your "no prefetch flag" commit?  Why metadata read for prefetch purposes is not a prefetch?  Why do we have statistic counters for metadata prefetch if almost none of metadata reads are counted as prefetch?

**Comment by ahrens** [comment]
@amotin That's right, the flag doesn't work.  I've filed a bug with a more detailed description of the issue.  I'm not sure of the right fix so I just removed the flag to keep the behavior the same for now.

https://www.illumos.org/issues/6852

**Comment by ahrens** [comment]
@pcd1193182 could you take a look at this when you get a chance?

File: usr/src/uts/common/fs/zfs/dmu_zfetch.c [hunk 1]
```diff
@@ -248,35 +259,74 @@ dmu_zfetch(zfetch_t *zf, uint64_t blkid, uint64_t nblks)
...
 
 	zs->zs_pf_blkid = pf_start + pf_nblks;
-	zs->zs_atime = gethrtime();
-	zs->zs_blkid = blkid + nblks;
 
 	/*
-	 * dbuf_prefetch() issues the prefetch i/o
-	 * asynchronously, but it may need to wait for an
-	 * indirect block to be read from disk.  Therefore
-	 * we do not want to hold any locks while we call it.
```
Comment (pcd1193182) [inline usr/src/uts/common/fs/zfs/dmu_zfetch.c]: Given that this block comment is no longer true, is there any redesigning we want to do here?

**Comment by pcd1193182** [comment]
LGTM

**Pull request closed without merging**